src/libstd/sys/unix/fd.rs

```diff
@@ -51,12 +51,20 @@ impl FileDesc {
         Ok(ret as usize)
     }
 
+    #[cfg(not(target_env = "newlib"))]
     pub fn set_cloexec(&self) {
         unsafe {
             let ret = c::ioctl(self.fd, c::FIOCLEX);
             debug_assert_eq!(ret, 0);
         }
     }
+    #[cfg(target_env = "newlib")]
+    pub fn set_cloexec(&self) {
+        unsafe {
+            let ret = c::fnctl(self.fd, c::F_SETFD, c::FD_CLOEXEC);
```

Comment:
Well, no *JS* doesn't have forking, but PNaCl does with help of `nacl-spawn` + JS code to insert additional `<embed>` elements (+ setting up pipes). However, my goal here was to support future use of Newlib on other platforms, as this issue isn't specific to NaCl (ie if some other libc was ported to use NaCl IRT).

Wrt `F_GETFD`: Yes I also had that thought, but considering such a sequence of operations wouldn't be thread safe in general without other serious design considerations (ie locks (which couldn't really be safe outside of Rust anyway)) and considering that `std::sys` is private to `std` anyway, and there aren't other flags to overwrite (unless such is added later, but then if no known flags could ever be added why would this patch try to address such?). So, why read the flags first place? Plus, the close-on-exec flags is the only flag defined.

Le sigh. Alright, I'll take what (I assume) would be your counterargument: just do it to save da future from pain.